pkg/kubelet/stats/cri_stats_provider.go

```diff
@@ -57,10 +65,213 @@ func newCRIStatsProvider(
+	// used as a cache to avoid querying cAdvisor for the filesystem stats with
+	// the same UUID many times.
+	uuidToFsInfo := make(map[runtimeapi.StorageIdentifier]*cadvisorapiv2.FsInfo)
+
+	// sandboxIDToPodStats is a temporary map from sandbox ID to its pod stats.
+	sandboxIDToPodStats := make(map[string]*statsapi.PodStats)
+
+	resp, err := p.runtimeService.ListContainerStats(&runtimeapi.ListContainerStatsRequest{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all container stats: %v", err)
+	}
+	for _, stats := range resp.Stats {
+		containerID := stats.Attributes.Id
+		container, found := containerMap[containerID]
+		if !found {
+			glog.Errorf("Unknown id %q in container map: %v", containerID)
+			continue
+		}
+
+		podSandboxID := container.PodSandboxId
+		podSandbox, found := podSandboxMap[podSandboxID]
+		if !found {
+			glog.Errorf("Unknown id %q in pod sandbox map: %v", podSandboxID)
+			continue
+		}
+
+		// Creates the stats of the pod (if not created yet) which the
+		// container belongs to.
+		ps, found := sandboxIDToPodStats[podSandboxID]
+		if !found {
+			ps = p.makePodStats(podSandbox)
+			sandboxIDToPodStats[podSandboxID] = ps
+		}
+		ps.Containers = append(ps.Containers, *p.makeContainerStats(stats, container, &rootFsInfo, uuidToFsInfo))
+	}
+
+	result := make([]statsapi.PodStats, 0, len(sandboxIDToPodStats))
+	for _, s := range sandboxIDToPodStats {
+		result = append(result, *s)
+	}
+	return result, nil
 }
 
+// ImageFsStats returns the stats of the image filesystem.
 func (p *criStatsProvider) ImageFsStats() (*statsapi.FsStats, error) {
-	return nil, fmt.Errorf("not implemented")
+	resp, err := p.imageService.ImageFsInfo(&runtimeapi.ImageFsInfoRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	// CRI may return the stats of multiple image filesystems but we only
+	// return the first one.
+	for _, fs := range resp.ImageFilesystems {
+		s := &statsapi.FsStats{
+			Time:       metav1.NewTime(time.Unix(0, fs.Timestamp)),
+			UsedBytes:  &fs.UsedBytes.Value,
+			InodesUsed: &fs.InodesUsed.Value,
+		}
+		imageFsInfo := p.getFsInfo(fs.StorageId)
+		if imageFsInfo != nil {
+			// The image filesystem UUID is unknown to the local node or
+			// there's an error on retrieving the stats. In these cases, we
+			// omit those stats and return the best-effort partial result. See
+			// https://github.com/kubernetes/heapster/issues/1793.
+			s.AvailableBytes = &imageFsInfo.Available
+			s.CapacityBytes = &imageFsInfo.Capacity
+			s.InodesFree = imageFsInfo.InodesFree
+			s.Inodes = imageFsInfo.Inodes
+		}
+		return s, nil
+	}
+
+	return nil, fmt.Errorf("imageFs information is unavailable")
+}
+
+// getFsInfo returns the information of the filesystem with the specified
+// storageID. If any error occurs, this function logs the error and returns
+// nil.
+func (p *criStatsProvider) getFsInfo(storageID *runtimeapi.StorageIdentifier) *cadvisorapiv2.FsInfo {
```

Comment:
Why don't we return error, and let the caller to decide whether to log error or not?

Reply:
In this way, we want to have the logging logic in a shared function.